developer-hub/src/reference_docs/extrinsic_encoding.rs

```diff
@@ -0,0 +1,172 @@
+//! # Constructing and Signing Extrinsics
+//!
+//! Extrinsics are payloads that are stored in blocks which are responsible for altering the state
+//! of a blockchain via the [_state transition function_][crate::reference_docs::blockchain_state_machines].
+//!
+//! Substrate is configurable enough that extrinsics can take any format. In practice, runtimes
+//! tend to use our [`sp_runtime::generic::UncheckedExtrinsic`] type to represent extrinsics,
+//! because it's generic enough to cater for most (if not all) use cases. In Polkadot, this is configured
+//! [here](https://github.com/polkadot-fellows/runtimes/blob/94b2798b69ba6779764e20a50f056e48db78ebef/relay/polkadot/src/lib.rs#L1478)
+//! at the time of writing.
+//!
+//! What follows is a description of how extrinsics based on this
+//! [`sp_runtime::generic::UncheckedExtrinsic`] type are encoded into bytes. Specifically, we are
+//! looking at how extrinsics with a format version of 4 are encoded. This version is itself a part
+//! of the payload, and if it changes, it indicates that something about the encoding may have
+//! changed.
+//!
+//! # Encoding an Extrinsic
+//!
+//! At a high level, all extrinsics compatible with [`sp_runtime::generic::UncheckedExtrinsic`]
+//! are formed from concatenating some details together, as in the following pseudo-code:
+//!
+//! ```text
+//! extrinsic_bytes = concat(
+//!     compact_encoded_length,
+//!     version_and_maybe_signature,
+//!     call_data
+//! )
+//! ```
+//!
+//! For clarity, the actual implementation
+//! [is here](https://github.com/paritytech/polkadot-sdk/blob/35eb133baab93d3f2f179df216b2cc175d7dcaf2/substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs#L296)
```

Comment:
Since we are in the mono-repo, I would 100% avoid using permalinks with commit hash. We can probably create a link to the `impl` header in rust-docs, no?

Reply:
Ah yeah; these are sortof hangovers from when it was a gist, but also I'm didn't think you could link in a nice way to impls etc in rust docs (only to definitions of things). Lemme see if there's a nice way!

If that doesn't work, I used permalinks with commit hashes just to guarantee that the link would never "break" (it may fall out of date, which is why I write "at the itme of writing" with these, but at least it won't end up pointing to complete nonsense if the code changes underneath :))

Reply:
I added the doc example instead for now!